File: RayVE.CSharp/Ray.cs
using RayVE.LinearAlgebra;

namespace RayVE
{
    public class Ray
    {
        public Point3D Origin { get; }

        public Vector3D Direction { get; }

        public Ray(Point3D origin, Vector3D direction)
        {
            Origin = origin;
            Direction = direction;
        }

        public Point3D GetPosition(double distance)
            => Origin + (distance * Direction);

        #region Operators

        public static Ray Multiply(Ray ray, Matrix matrix)
        {
            if (matrix.ColumnCount != 4)
                throw new DimensionMismatchException();

            var transformedOrigin = ray.Origin * matrix;
            var transformedDirection = ray.Direction * matrix;

            return new Ray(transformedOrigin, transformedDirection);
        }

        public static Ray operator *(Ray ray, Matrix matrix)
            => Multiply(ray, matrix);

        public static Ray Multiply(Matrix matrix, Ray ray)
        {
            if (matrix.ColumnCount != 4)
                throw new DimensionMismatchException();

            var transformedOrigin = matrix * ray.Origin;
            var transformedDirection = matrix * ray.Direction;

            return new Ray(transformedOrigin, transformedDirection);
        }

        public static Ray operator *(Matrix matrix, Ray ray)
            => Multiply(matrix, ray);
        #endregion
}
}

File: RayVE.FSharp/Ray.fs
namespace RayVE

type Ray (origin: Point3D, direction: Vector3D) =
    member __.Origin = origin
    member __.Direction = Vector3D.normalize direction

    static member (*) (ray: Ray, matrix: Matrix) =
        if matrix.ColumnCount <> 4 then
            invalidArg (nameof matrix) "Matrix must have 4 columns."
        (ray.Origin * matrix, ray.Direction * matrix) |> Ray

    static member (*) (matrix: Matrix, ray: Ray) =
        if matrix.ColumnCount <> 4 then
            invalidArg (nameof matrix) "Matrix must have 4 columns."
        (matrix * ray.Origin, matrix * ray.Direction) |> Ray
    
    override __.Equals (other: obj): bool = 
        match other with
        | :? Ray as r -> 
            Vector3D.magnitude(__.Origin - r.Origin) < Constants.EPSILON
            && Vector3D.magnitude(__.Direction - r.Direction) < Constants.EPSILON
        | _ -> false

    static member op_Equality (ray1: Ray, ray2: Ray) =
        Vector3D.magnitude(ray1.Origin - ray2.Origin) < Constants.EPSILON
        && Vector3D.magnitude(ray1.Direction - ray2.Direction) < Constants.EPSILON

module Ray =
    let position (ray: Ray) (distance: double) =
        ray.Origin + distance * ray.Direction